syntax = "proto3";

import "proto/transaction.proto";

package slog.api;

/***********************************************
                    REQUESTS
***********************************************/

message Request {
    uint32 stream_id = 1;
    oneof type {
        TransactionRequest txn = 2;
        StatsRequest stats = 3;
    }
}

message TransactionRequest {
    Transaction txn = 1;
}

enum StatsModule {
    SERVER = 0;
    FORWARDER = 1;
    MHORDERER = 2;
    SEQUENCER = 3;
    SCHEDULER = 4;
}

message StatsRequest {
    StatsModule module = 1;
    // Level of details, starting from 0
    uint32 level = 2;
}

/***********************************************
                    RESPONSES
***********************************************/

message Response {
    uint32 stream_id = 1;
    oneof type {
        TransactionResponse txn = 2;
        StatsResponse stats = 3;
    }
}

// For debugging and testing purposes
message TransactionResponse {
    Transaction txn = 1;
}

message StatsResponse {
    bytes stats_json = 1;
}